hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotDefragService.java

```diff
@@ -0,0 +1,1138 @@
+    String previousSnapshotId = currentSnapshot.getPathPreviousSnapshotId() != null ?
+        currentSnapshot.getPathPreviousSnapshotId().toString() : null;
+
+    while (previousSnapshotId != null) {
+      try (TableIterator<String, ? extends Table.KeyValue<String, SnapshotInfo>> iterator =
+               snapshotInfoTable.iterator()) {
+        iterator.seekToFirst();
+
+        while (iterator.hasNext()) {
+          Table.KeyValue<String, SnapshotInfo> keyValue = iterator.next();
+          SnapshotInfo snapshotInfo = keyValue.getValue();
+
+          if (snapshotInfo.getSnapshotId().toString().equals(previousSnapshotId)) {
+            if (!needsDefragmentation(snapshotInfo)) {
+              LOG.info("Found previous defragmented snapshot: {} (ID: {})",
+                  snapshotInfo.getName(), snapshotInfo.getSnapshotId());
+              return snapshotInfo;
+            }
+
+            // Continue searching with this snapshot's previous
+            previousSnapshotId = snapshotInfo.getPathPreviousSnapshotId() != null ?
+                snapshotInfo.getPathPreviousSnapshotId().toString() : null;
+            break;
+          }
+        }
+      }
+    }
+
+    LOG.debug("No previous defragmented snapshot found");
+    return null;
+  }
+
+  /**
+   * Performs full defragmentation for the first snapshot in the chain.
+   * This is a simplified implementation that demonstrates the concept.
+   */
+  private void performFullDefragmentation(SnapshotInfo snapshotInfo,
+      OmSnapshot omSnapshot) throws IOException {
+
+    String snapshotPath = OmSnapshotManager.getSnapshotPath(
+        ozoneManager.getConfiguration(), snapshotInfo);
+
+    // For defraggedDbPath, we need to go up to the parent directory and use checkpointStateDefragged
+    String parentDir = Paths.get(snapshotPath).getParent().getParent().toString();
+    String checkpointDirName = Paths.get(snapshotPath).getFileName().toString();
+    String defraggedDbPath = Paths.get(parentDir, CHECKPOINT_STATE_DEFRAGGED_DIR, checkpointDirName).toString();
+
+    LOG.info("Starting full defragmentation for snapshot: {} at path: {}",
+        snapshotInfo.getName(), snapshotPath);
+    LOG.info("Target defragmented DB path: {}", defraggedDbPath);
+
+    // Create defragmented directory
+    Files.createDirectories(Paths.get(defraggedDbPath));
+
+    // TODO: Get snapshot checkpoint DB via SnapshotCache
+    RDBStore originalStore = (RDBStore) omSnapshot.getMetadataManager().getStore();
+    RocksDatabase originalDb = originalStore.getDb();
+
+    LOG.info("Starting defragmentation process for snapshot: {}", snapshotInfo.getName());
+    LOG.info("Original DB path: {}", snapshotPath);
+    LOG.info("Defragmented DB path: {}", defraggedDbPath);
+
+    // Implement actual RocksDB defragmentation
+    try {
+      // 1. Create a new RocksDB instance at defraggedDbPath
+      DBStoreBuilder defraggedDbBuilder = DBStoreBuilder.newBuilder(ozoneManager.getConfiguration())
```

Comment:
To copy we can just take the current checkpoint and do force compaction of the entire db.
Step to be performed:

1. Take checkpoint of current DB.
2. DeleteRange KeyTable from ["", keyTablePrefix) + [lexicographicalHigherString(keyTablePrefix), lexicographicalHigherString("/")]
3. DeleteRange DirectoryTable from ["", FSOPrefix) + [lexicographicalHigherString(FSOPrefix), lexicographicalHigherString("/")]
4. DeleteRange FileTable from ["", FSOPrefix) + [lexicographicalHigherString(FSOPrefix), lexicographicalHigherString("/")]
5. kForceCompact all tables in the checkpoint.

Reply:
done